File: Assignment3/Solution/python/Xiaoyu_Zheng_task2_ItemBasedCF.py
from __future__ import division

from pyspark.sql import SparkSession
from itertools import combinations

import time
import sys

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("""Usage:
             cd $SPARK_HOME
             bin/spark-submit Xiaoyu_Zheng_task2_ItemBasedCF.py <rating file path> <testing file path>
          """)
        exit(-1)

    path = sys.argv[1]
    test = sys.argv[2]
    start = time.time()
    # path = "/Users/xiaoyuzheng/Desktop/usc/18summer/553/Assignments/Assignment_03/data/ml-latest-small/ratings.csv"
    # path = "/Users/xiaoyuzheng/Desktop/usc/18summer/553/Assignments/Assignment_03/data/ml-20m/ratings.csv"
    # print(path)
    # test = "/Users/xisaoyuzheng/Desktop/usc/18summer/553/Assignments/Assignment_03/data/testing_small.csv"
    # test = "/Users/xiaoyuzheng/Desktop/usc/18summer/553/Assignments/Assignment_03/data/testing_20m.csv"

    spark = SparkSession.builder \
        .appName("ItemBased") \
        .getOrCreate()

    sc = spark.sparkContext
    rdd = sc.textFile(path)
    header = rdd.first()
    data = rdd.filter(lambda row: row != header).map(lambda x: x.split(","))

    mov = data.map(lambda x: (int(x[1]), 1)).reduceByKey(lambda x, y: x).map(lambda x: x[0])
    movies = list(mov.collect())
    movies.sort()
    dicM = {}
    for i, e in enumerate(movies):
        dicM[e] = i
    usr = data.map(lambda x: (int(x[0]), 1)).reduceByKey(lambda x, y: x).map(lambda x: x[0])
    users = list(usr.collect())
    users.sort()
    dicU = {}
    for i, e in enumerate(users):
        dicU[e] = i
    vu = sc.broadcast(dicU)

    mat = data.map(lambda x: (int(x[1]), [vu.value[int(x[0])]])).reduceByKey(lambda x, y: x+y).sortBy(lambda x: x[0])
    matrix = mat.collect()

    # print(matrix[dicM[135536]])
    # print(matrix[dicM[96114]])

    m = len(users)  # m: the number of the bins

    # hash function:
    def f(x, has):
        a = has[0]
        b = has[1]
        p = has[2]
        # return ((a * x + b) % p) % m
        return min([((a * e + b) % p) % m for e in x[1]])

    # TODO: more hash functions (or less?)
    # hash parameters [a, b, p]
    hashes = [[913, 901, 24593], [14, 23, 769], [1, 101, 193], [17, 91, 1543], \
              [387, 552, 98317], [11, 37, 3079], [2, 63, 97], [41, 67, 6151], \
              [91, 29, 12289], [3, 79, 53], [73, 803, 49157], [8, 119, 389]]
    # good hash table primes: http://planetmath.org/goodhashtableprimes

    # def hashMatrix(x):
    #     return [f(x, has) for has in hashes]

    # print(matrix[0])
    # print(matrix[0][0], matrix[1][0], matrix[2][0], matrix[100][0])
    # print(movies[0], movies[1], movies[2], movies[100])

    signatures = mat.map(lambda x: (x[0], [f(x, has) for has in hashes]))
    # print(signatures.collect()[:10])

    n = len(hashes)  # the size of the signature column
    b = 6
    r = int(n/b)

    def sig(x):
        # for e in x:
        res = []
        for i in range(b):
            res.append(((i, tuple(x[1][i*r:(i+1)*r])), [x[0]]))
        return res

    def pairs(x):
        res = []
        length = len(x[1])
        whole = list(x[1])
        whole.sort()
        for i in range(length):
            for j in range(i+1, length):
                res.append(((whole[i], whole[j]), 1))
        return res

    # cand = signatures.flatMap(sig)
    cand = signatures.flatMap(sig).reduceByKey(lambda x, y: x+y).filter(lambda x: len(x[1]) > 1).flatMap(pairs) \
        .reduceByKey(lambda x,y: x).map(lambda x: x[0])

    # for e in cand.collect():
    #     print(e)

    # sign = signatures.collect()
    # size = len(movies)
    # candidate = []
    # for i in range(size):
    #     for j in range(i+1, size):
    #         piece1 = sign[i]
    #         piece2 = sign[j]
    #         for k in range(b):
    #             if piece1[1][k*r:(k+1)*r] == piece2[1][k*r:(k+1)*r]:
    #                 candidate.append([piece1[0], piece2[0]])
    #                 break
    # print("candidate pairs:")
    # print(candidate)

    def jaccard(x):
        a = set(matrix[dicM[x[0]]][1])
        b = set(matrix[dicM[x[1]]][1])
        inter = a & b
        union = a | b
        jacc = len(inter)/len(union)
        return (x[0], x[1], jacc)

    # result = cand.map(jaccard).filter(lambda x: x[2] >= 0.5).sortBy(lambda x: x[1]).sortBy(lambda x: x[0])
    # print(result.take(10))

    # similar movie pairs from LSH: (movie1_id, movie2_id)
    simiMovie = cand.map(jaccard).filter(lambda x: x[2] >= 0.5)\
        .map(lambda x: ((min(x[0], x[1]), max(x[0], x[1])), x[2]))

    #     .reduceByKey(lambda x, y: x + y)
    # print(simiMovie.take(10))

    dataAll = data
    data = dataAll.map(lambda x: ((int(x[0]), int(x[1])), float(x[2])))
    testRdd = sc.textFile(test).map(lambda x: x.split(","))
    testHeader = testRdd.first()
    testKey = testRdd.filter(lambda row: row != testHeader).map(lambda x: (int(x[0]), int(x[1])))
    testData = testKey.map(lambda x: (x, None))

    # ((user_id, movie_id), rating)
    trainData = data.subtractByKey(testData)
    # print(trainData.take(10))
    dicTrain = {}
    for e in trainData.collect():
        dicTrain[e[0]] = e[1]

    # (movie_id, average_rating)
    movieAvg = trainData.map(lambda x: (x[0][1], [x[1]])).reduceByKey(lambda x, y: x+y).map(lambda x: (x[0], sum(x[1])/len(x[1])))
    dicAvg = {}
    for e in movieAvg.collect():
        dicAvg[e[0]] = e[1]
    # print(movieAvg.take(10))

    # (user_id, [(movie_id, rating)])
    userToMovie = trainData.map(lambda x: (x[0][0], [(x[0][1], x[1])])).reduceByKey(lambda x, y: x+y)
    # print(userToMovie.take(10))

    # =====================================================================================================
    def movieRatingComb(x):
        for y in combinations(x[1], 2):
            a, b = min(y[0], y[1]), max(y[0], y[1])
            yield ((a[0], b[0]), [(a[1], b[1])])

    # ((movie1_id, movie2_id), [(rating1, rating2)])    movie1_id < movie2_id
    movieToRating = userToMovie.flatMap(movieRatingComb).reduceByKey(lambda x, y: x+y)
    # print(movieToRating.take(10))


    # simiMovieToRating = movieToRating.filter(lambda x: x[0] in simiMovie)
    simiMovieToRating = movieToRating.join(simiMovie.map(lambda x: (x, None))).map(lambda x: (x[0], x[1][:-1]))
    # print("simiMovieToRating: ")
    # print(simiMovieToRating.take(10))

    def calSimi(x):
        ratings = x
        ratingA = [x[0] for x in ratings]
        ratingB = [x[1] for x in ratings]
        avg1 = sum(ratingA)/len(ratingA)
        avg2 = sum(ratingB)/len(ratingB)
        diff1 = [x-avg1 for x in ratingA]
        diff2 = [x-avg2 for x in ratingB]
        root1 = pow(sum([x**2 for x in diff1]), 0.5)
        root2 = pow(sum([x**2 for x in diff2]), 0.5)
        up = sum([diff1[i]*diff2[i] for i in range(len(ratingA))])
        if up == 0:
            return 0
        return up/(root1*root2)

    # ((movie1_id, movie2_id), pearson_similarity) movie1_id < movie2_id
    movieSimi = simiMovieToRating.map(lambda x: (x[0], calSimi(x[1])))
    # =====================================================================================================

    # =====================================================================================================
    # def movieRatingComb(x):
    #     for y in combinations(x[1], 2):
    #         a, b = min(y[0], y[1]), max(y[0], y[1])
    #         yield ((a[0], b[0]), None)
    # # ((movie1_id, movie2_id), None)
    # movieToRating = userToMovie.flatMap(movieRatingComb)
    # # ((movie1_id, movie2_id), similarity from LSH) ???????????????????????????????/
    # movieSimi = movieToRating.join(simiMovie).map(lambda x: (x[0], x[1][1]))
    # print(movieSimi.take(10))
    # =====================================================================================================

    movieSimiMatrix = movieSimi.flatMap(lambda x: [(x[0][0], [(x[0][1], x[1])]), (x[0][1], [(x[0][0], x[1])])]) \
        .reduceByKey(lambda x, y: x+y)

    dicSimi = {}
    for e in movieSimiMatrix.collect():
        dicSimi[e[0]] = e[1]

    def predict(x):
        if x[1] not in dicSimi:
            if x[1] not in dicAvg:
                return (x, 2.5)
            return (x, dicAvg[x[1]])
        allCom = dicSimi[x[1]]
        allCom.sort(key=lambda y: y[1], reverse=True)
        length = len(allCom)
        companion = []
        if length > 2:
            companion = allCom[:2]
        else:
            companion = allCom
        avg = 0
        sumi = 0
        down = 0
        if len(companion) == 0:
            print("no companion")
        for e in companion:
            tmp = (x[0], e[0])
            if tmp in dicTrain:
                rate = dicTrain[tmp]
                sumi += rate*e[1]
                down += abs(e[1])
        if down == 0:
            return (x, avg)
        return (x, sumi/down+avg)

    def cutPred(x):
        # rate = 0
        if x[1] > 5:
            rate = 5
        elif x[1] < 0:
            rate = 0
        else:
            rate = x[1]
        return (x[0], rate)

    result = testKey.map(predict).map(cutPred).sortBy(lambda x: x[0][1]).sortBy(lambda x: x[0][0])
    print(result.take(10))

    ratesAndPreds = data.join(testData).map(lambda x: (x[0], x[1][0])).join(result)
    # ratesAndPreds = data.join(testKey).map(lambda x: (x[0], x[1][0])).join(result)
    diff = ratesAndPreds.map(lambda r: abs(r[1][0] - r[1][1]))
    diff01 = diff.filter(lambda x: 0 <= x < 1)
    diff12 = diff.filter(lambda x: 1 <= x < 2)
    diff23 = diff.filter(lambda x: 2 <= x < 3)
    diff34 = diff.filter(lambda x: 3 <= x < 4)
    diff4 = diff.filter(lambda x: 4 <= x)

    MSE = diff.map(lambda x: x**2).mean()
    RMSE = pow(MSE, 0.5)

    text = open("Xiaoyu_Zheng_ItemBasedCF.txt", "w")
    if result:
        for e in result.collect():
            text.write(str(e[0][0]))
            text.write(", ")
            text.write(str(e[0][1]))
            text.write(", ")
            text.write(str(e[1]))
            text.write("\n")
    text.close()

    print(">=0 and <1: "+str(diff01.count()))
    print(">=1 and <2: "+str(diff12.count()))
    print(">=2 and <3: "+str(diff23.count()))
    print(">=3 and <4: "+str(diff34.count()))
    print(">=4: "+str(diff4.count()))
    print("RMSE: "+str(RMSE))

    end = time.time()
    print("Time: "+str(end - start)+" sec")
    spark.stop()

File: Assignment3/Solution/scala/Xiaoyu_Zheng_task2_ItemBasedCF.scala
import java.io.{BufferedWriter, File, FileWriter}

import breeze.linalg.sum
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object ItemBasedCF {
  def main(args: Array[String]): Unit = {
    val start = System.nanoTime()
        if (args.length != 2) {
          println(
            """
              |Usage:
              |cd $SPARK_HOME
              |bin/spark-submit --class ItemBasedCF Xiaoyu_Zheng_hw3.jar <rating file path> <testing file path>
            """.stripMargin)
          System.exit(-1)
        }
        val path = args(0)
        val test = args(1)
//    val path = "/Users/xiaoyuzheng/Desktop/usc/18summer/553/Assignments/Assignment_03/data/ml-latest-small/ratings.csv"
//    val test = "/Users/xiaoyuzheng/Desktop/usc/18summer/553/Assignments/Assignment_03/data/testing_small.csv"
    val conf = new SparkConf()
    conf.setAppName("ItemBasedCF-Scala")
    conf.setMaster("local[4]")
    val sc = new SparkContext(conf)
    val rdd = sc.textFile(path)
    val header = rdd.first()
    val data = rdd.filter(row => row != header).map(x => x.split(","))

    val mov = data.map(x => (x(1).toInt, 1)).reduceByKey((x, y) => x).map(x => x._1)
    //    val mov = movi.map(x => x(0))
    // list of all movie_id
    val movies = mov.collect().toList.sorted
    val dicM = scala.collection.mutable.HashMap.empty[Int, Int]
    var indexOfMovies = 0
    for (e <- movies) {
      dicM(e) = indexOfMovies
      indexOfMovies += 1
    }
    val usr = data.map(x => (x(0).toInt, 1)).reduceByKey((x, y) => x).map(x => x._1)
    //    val mov = movi.map(x => x(0))
    // list of all movie_id
    val users = usr.collect().toList.sorted
    val dicU = scala.collection.mutable.HashMap.empty[Int, Int]
    var indexOfUsers = 0
    for (e <- users) {
      dicU(e) = indexOfUsers
      indexOfUsers += 1
    }
    //    print(dicU)
    val vu = sc.broadcast(dicU)
    val mat = data.map(x => {
      val tmp = scala.collection.mutable.ListBuffer.empty[Int]
      tmp += vu.value(x(0).toInt)
      Tuple2(x(1).toInt, tmp.toList)
    })
      .reduceByKey((x, y) => List.concat(x, y))
      .sortBy(x => x._1)
    val matrix = mat.collect()

    val m = users.length  // m: the number of the bins


    //    print(matrix)
    // hash function
    //    def f(x: Iterator[])
    def f(x: (Int, List[Int]), has: List[Int]): Int = {
      val a = has(0)
      val b = has(1)
      val p = has(2)
      val tmp = scala.collection.mutable.ListBuffer.empty[Int]
      for (e <- x._2) {
        val temp = ((a * e + b) % p) % m
        tmp += temp
      }
      val length = tmp.length
      var min = Int.MaxValue
      if (length > 0) {
        for (e <- tmp) {
          min = Math.min(min, e)
        }
      }
      min
    }

    val hashes = List(List(913, 901, 24593), List(14, 23, 769), List(1, 101, 193),
      List(17, 91, 1543), List(387, 552, 98317), List(11, 37, 3079), List(2, 63, 97),
      List(41, 67, 6151), List(91, 29, 12289), List(3, 79, 53), List(73, 803, 49157),
      List(8, 119, 389))

    // (movie_id, [signature])
    val signatures = mat.map(x => {
      //      var tmp = 0
      val tmp = scala.collection.mutable.ListBuffer.empty[Int]
      for (has <- hashes) {
        tmp += f(x, has)
      }
      Tuple2(x._1, tmp.toList)
    })

    //    val sig = signatures.collect()

    //    for (i <- 0 to 10) {
    //      print(sig(i))
    //    }

    //    print(signatures.collect().take(10))

    val n = hashes.length // 12
    val b = 6
    val r = n/b

    def sig(x: (Int, List[Int])): List[((Int, Tuple1[List[Int]]), List[Int])] = {
      //      val res = scala.collection.mutable.ListBuffer.empty[Tuple2[Tuple2[Int, Tuple1[List[Int]]], List[Int]]]
      val res = scala.collection.mutable.ListBuffer.empty[((Int, Tuple1[List[Int]]), List[Int])]
      for (i <- 0 until b) {
        res += Tuple2(Tuple2(i, Tuple1(x._2.slice(i*r, (i+1)*r))), List(x._1))
      }
      res.toList
    }

    def pairs(x: ((Int, Tuple1[List[Int]]), List[Int])): List[((Int, Int), Int)] = {
      val res = scala.collection.mutable.ListBuffer.empty[((Int, Int), Int)]
      val length = x._2.length
      val whole = x._2.sorted
      for (i <- 0 until length) {
        for (j <- i+1 until length) {
          res += Tuple2(Tuple2(whole(i), whole(j)), 1)
        }
      }
      res.toList
    }

    // [(movie1_id, movie2_id)]
    //    val candPre1: RDD[((Int, Tuple1[List[Int]]), List[Int])] = signatures.flatMap(x => sig(x)).reduceByKey((x, y) => x ++ y)
    //                                                                        .filter(x => x._2.length > 1)
    //    val candPre2: RDD[((Int, Int), Int)] = candPre1.flatMap(x => pairs(x)).reduceByKey((x, y) => x)
    //    val cand : RDD[(Int, Int)] = candPre2.map(x => x._1)
    val cand = signatures.flatMap(x => sig(x)).reduceByKey((x, y) => x ++ y)
      .filter(x => x._2.length > 1).flatMap(x => pairs(x)).reduceByKey((x, y) => x).map(x => x._1)

    def jaccard(x: (Int, Int)): (Int, Int, Float) = {
      val a = matrix(dicM(x._1))._2.toSet
      val b = matrix(dicM(x._2))._2.toSet
      val inter = a.intersect(b)
      val union = a.union(b)
      val jacc = inter.size.toFloat/union.size
      Tuple3(x._1, x._2, jacc)
    }
//    val result = cand.map(x => jaccard(x)).filter(x => x._3 >= 0.5).sortBy(x => x._2).sortBy(x => x._1)
//    println(result.take(10))

    val simiMovie = cand.map(x => jaccard(x)).filter(x => x._3 >= 0.5)
      .map(x => ((Math.min(x._1, x._2), Math.max(x._1, x._2)), x._3))

    val dataAll = data.map(x => ((x(0).toInt, x(1).toInt), x(2).toFloat))
    val testRdd = sc.textFile(test)
    val testHeader = testRdd.first()
    val testKey = testRdd.filter(row => row != testHeader).map(x => x.split(",")).map(x => (x(0).toInt, x(1).toInt))
    val testData = testKey.map(x => (x, 0.toFloat))

    // ((user_id, movie_id), rating)
    val trainData = dataAll.subtractByKey(testData)

    // hashmap(key:(user_id, movie_id), value: rating)
    val dicTrain = scala.collection.mutable.HashMap.empty[(Int, Int), Float]
    for (e <- trainData.collect()) {
      dicTrain(e._1) = e._2
    }

    // (movie_id, rating_avg)
    val movieAvg = trainData.map(x => {
      val tmp = scala.collection.mutable.ListBuffer.empty[Float]
      tmp += x._2
      (x._1._2, tmp.toList)
    }).reduceByKey((x, y) => x ++ y).map(x => (x._1, sum(x._2)/x._2.length))

    // hashmap(key:movie_id, value: rating)
    val dicAvg = scala.collection.mutable.HashMap.empty[Int, Float]
    for (e <- movieAvg.collect()) {
      dicAvg(e._1) = e._2
    }

    // (user_id, [(movie_id, rating)])
    val userToMovie = trainData.map(x => {
      val tmp = scala.collection.mutable.ListBuffer.empty[(Int, Float)]
      tmp += Tuple2(x._1._2, x._2)
      (x._1._1, tmp.toList)
    }).reduceByKey((x, y) => x++y)

    def movieRatingComb(x: (Int, List[(Int, Float)])): List[((Int, Int), List[(Float, Float)])] = {
      val tmp = scala.collection.mutable.ListBuffer.empty[((Int, Int), List[(Float, Float)])]
      for (y <- x._2.combinations(2)) {
        val temp = scala.collection.mutable.ListBuffer.empty[(Float, Float)]
        var a = (0, 0.toFloat)
        var b = (0, 0.toFloat)
        if (y(0)._1 > y(1)._1) {
          a = y(1)
          b = y(0)
        }
        else {
          a = y(0)
          b = y(1)
        }
        temp += Tuple2(a._2, b._2)
        tmp += Tuple2((a._1, b._1), temp.toList)
      }
      tmp.toList
    }
    // ((movie1_id, movie2_id), [(rating1, rating2)])    movie1_id < movie2_id
    val movieToRating = userToMovie.flatMap(movieRatingComb).reduceByKey((x, y) => x++y)

    val simiMovieToRating: RDD[((Int, Int), List[(Float, Float)])] = movieToRating.join(simiMovie).map(x => (x._1, x._2._1.slice(0, -1)))

    def calSimi(x: List[(Float, Float)]): Float = {
      var ratingA = scala.collection.mutable.ListBuffer.empty[Float]
      var ratingB = scala.collection.mutable.ListBuffer.empty[Float]
      //      var diff1 = scala.collection.mutable.ListBuffer.empty[Float]
      //      var diff2 = scala.collection.mutable.ListBuffer.empty[Float]
      var diff1 = 0.toFloat
      var diff2 = 0.toFloat
      var up = 0.toFloat
      //      var downSq1 = scala.collection.mutable.ListBuffer.empty[Float]
      //      var downSq2 = scala.collection.mutable.ListBuffer.empty[Float]
      var downSq1 = 0.toFloat
      var downSq2 = 0.toFloat
      for (e <- x) {
        ratingA += e._1
        ratingB += e._2
      }
      val avg1 = sum(ratingA)/ratingA.length
      val avg2 = sum(ratingB)/ratingB.length
      for (i <- ratingA.indices) {
        diff1 = ratingA(i) - avg1
        diff2 = ratingA(i) - avg1
        downSq1 += Math.pow(diff1, 2).toFloat
        downSq2 += Math.pow(diff2, 2).toFloat
        up += diff1 * diff2
      }
      val root1 = Math.pow(downSq1, 0.5)
      val root2 = Math.pow(downSq2, 0.5)
      val down = root1 * root2
      if (up == 0) {
        return 0
      }
      (up/down).toFloat
    }

    // ((movie1_id, movie2_id), pearson_similarity) movie1_id < movie2_id
    val movieSimi = simiMovieToRating.map(x => (x._1, calSimi(x._2)))

    // (activeMovie_id, [(movie1_id, similarity)])
    val movieSimiMatrix = movieSimi.flatMap(x => {
      var tmp = scala.collection.mutable.ListBuffer.empty[(Int, List[(Int, Float)])]
      var temp1 = scala.collection.mutable.ListBuffer.empty[(Int, Float)]
      var temp2 = scala.collection.mutable.ListBuffer.empty[(Int, Float)]
      temp1 += Tuple2(x._1._2, x._2)
      temp2 += Tuple2(x._1._1, x._2)
      tmp += Tuple2(x._1._1, temp1.toList)
      tmp += Tuple2(x._1._2, temp2.toList)
      tmp.toList
    }).reduceByKey((x, y) => x ++ y)

    val dicSimi = scala.collection.mutable.HashMap.empty[Int, List[(Int, Float)]]
    for (e <- movieSimiMatrix.collect()) {
      dicSimi(e._1) = e._2
    }

    def predict(x: (Int, Int)): ((Int, Int), Float) = {
      if (!dicSimi.contains(x._2)) {
        if (!dicAvg.contains(x._2)) {
          return (x, 2.5.toFloat)
        }
        return (x, dicAvg(x._2))
      }
      val allCom = dicSimi(x._2)
      //      allCom.sorted()
      val length = allCom.size
      //      var companion = List()
      //      var companion = scala.collection.mutable.ListBuffer.empty[(Int, Float)].toList
      var companion = List[(Int, Float)]()
      if (length > 5) {
        companion = allCom.slice(0, 5)
      }
      else {
        companion = allCom
      }
      val avg = 0.toFloat
      var sumi = 0.toFloat
      var down = 0.toFloat
      //      if (companion)
      for (e <- companion) {
        val tmp = Tuple2(x._1, e._1)
        if (dicTrain.contains(tmp)) {
          val rate = dicTrain(tmp)
          sumi += rate * e._2
          down += Math.abs(e._2)
        }
      }
      var returns = Tuple2(Tuple2(0, 0), 0.toFloat)
      if (down == 0) {
        returns = Tuple2(x, dicAvg(x._2))
      }
      else {
        returns = Tuple2(x, sumi/down+avg)
      }
      returns
    }

    def cutPred(x: ((Int, Int), Float)): ((Int, Int), Float) = {
      var rate = 0.toFloat
      if (x._2 > 5) {
        rate = 5
      }
      else if (x._2 < 1) {
        rate = 1
      }
      else {
        rate = x._2
      }
      (x._1, rate)
    }

    val predictions = testKey.map(x => predict(x)).map(cutPred).sortBy(x => x._1._2).sortBy(x => x._1._1)

    val ratesAndPreds = dataAll.join(testData).map(x => (x._1, x._2._1)).join(predictions)
    val diff = ratesAndPreds.map(r => Math.abs(r._2._1 - r._2._2))
    val diff01 = diff.filter(x => x >= 0 && x < 1).count()
    val diff12 = diff.filter(x => x >= 1 && x < 2).count()
    val diff23 = diff.filter(x => x >= 2 && x < 3).count()
    val diff34 = diff.filter(x => x >= 3 && x < 4).count()
    val diff4 = diff.filter(x => x >= 4).count()
    val MSE = diff.map(x => Math.pow(x, 2)).mean()
    val RMSE = Math.pow(MSE, 0.5)

    val file = new File("Xiaoyu_Zheng_ItemBasedCF.txt")
    val bw = new BufferedWriter(new FileWriter(file))
    val it = predictions.collect().iterator
    while (it.hasNext) {
      var e = it.next()
      //      var size = point.length
      bw.write(e._1._1.toString)
      bw.write(", ")
      bw.write(e._1._2.toString)
      bw.write(", ")
      bw.write(e._2.toString)
      bw.write("\n")
    }
    bw.close()
    println(">=0 and <1: "+diff01.toString)
    println(">=1 and <2: "+diff12.toString)
    println(">=2 and <3: "+diff23.toString)
    println(">=3 and <4: "+diff34.toString)
    println(">=4: "+diff4.toString)
    println("RMSE: "+RMSE.toString)
    val end = System.nanoTime()
    println("Time: "+((end-start)/1e9d).toString+" sec")
  }

}
